llvm/lib/Target/AIE/AIE2InstructionSelector.cpp

```diff
@@ -4140,6 +4266,168 @@ bool AIE2InstructionSelector::select512BitG_AIE_STORE_SRS(
   }
 }
 
+// Returns the instruction where the Reg is defined ignoring G_BITCAST.
+static MachineInstr *stripBitCast(Register Reg, MachineRegisterInfo &MRI) {
+  MachineInstr *MI = getDefIgnoringCopies(Reg, MRI);
+  if (MI->getOpcode() == TargetOpcode::G_BITCAST)
+    return getDefIgnoringCopies(MI->getOperand(1).getReg(), MRI);
+  return MI;
+}
+
+// If two 256-bit SRS are input to 512-bit STORE through CONCAT, these are
+// combined into VST.SRS
+bool AIE2InstructionSelector::selectG_AIE_STORE_CONCAT_SRS(
+    MachineInstr &StoreI, MachineRegisterInfo &MRI) {
+
+  Register ConcatResult = (StoreI.uses().begin())->getReg();
+  MachineInstr *ConcatOp = MRI.getUniqueVRegDef(ConcatResult);
+
+  assert(ConcatOp && "Expected SSA.");
+
+  if (!canDelayMemOp(*ConcatOp, StoreI, MRI))
+    return false;
+
+  if (getLoadStoreSize(StoreI) != 512 ||
+      ConcatOp->getOpcode() != AIE2::G_INTRINSIC ||
+      cast<GIntrinsic>(*ConcatOp).getIntrinsicID() !=
+          Intrinsic::aie2_concat_I512_I256)
+    return false;
+
+  MachineInstr *SrsOpLo = stripBitCast(ConcatOp->getOperand(2).getReg(), MRI);
+  MachineInstr *SrsOpHi = stripBitCast(ConcatOp->getOperand(3).getReg(), MRI);
+
+  if (!canCombineSRSUPS(StoreI, *SrsOpLo) ||
+      !canCombineSRSUPS(StoreI, *SrsOpHi))
+    return false;
+
+  std::optional<AddressingModeInfo> AMI =
+      getOrDefineAddressingRegister(StoreI, MRI);
+  if (!AMI)
+    return false;
+
+  auto [SrcRegLo, ShftRegLo, SignRegLo] = std::tuple(
+      SrsOpLo->getOperand(2).getReg(), SrsOpLo->getOperand(3).getReg(),
+      SrsOpLo->getOperand(4).getReg());
+  auto [SrcRegHi, ShftRegHi, SignRegHi] = std::tuple(
+      SrsOpHi->getOperand(2).getReg(), SrsOpHi->getOperand(3).getReg(),
+      SrsOpHi->getOperand(4).getReg());
+
+  std::optional<LoadStoreOpcodes> LSOLo;
+  bool ConstantSignLo = false;
+  if (auto SignVal = getIConstantVRegValWithLookThrough(SignRegLo, MRI)) {
+    // SignVal = 1 for signed and 0 for dynamically signed
+    LSOLo = getCombinedOpcodeSRSUPS(StoreI, *SrsOpLo, AMI->ImmediateOffset,
+                                    SignVal.value().Value == 0x1);
+    ConstantSignLo = true;
+  } else {
+    LSOLo =
+        getCombinedOpcodeSRSUPS(StoreI, *SrsOpLo, AMI->ImmediateOffset, false);
+  }
+
+  assert(LSOLo && "Unexpected VST.SRS combine failure");
+
+  std::optional<LoadStoreOpcodes> LSOHi;
+  bool ConstantSignHi = false;
+  if (auto SignVal = getIConstantVRegValWithLookThrough(SignRegHi, MRI)) {
+    // SignVal = 1 for signed and 0 for dynamically signed
+    LSOHi = getCombinedOpcodeSRSUPS(StoreI, *SrsOpHi, AMI->ImmediateOffset,
+                                    SignVal.value().Value == 0x1);
+    ConstantSignHi = true;
+  } else {
+    LSOHi =
+        getCombinedOpcodeSRSUPS(StoreI, *SrsOpHi, AMI->ImmediateOffset, false);
+  }
+
+  assert(LSOHi && "Unexpected VST.SRS combine failure");
+
+  switch (AMI->MemI.getOpcode()) {
```

Comment:
That insertion logic here is very similar to `select512BitG_AIE_STORE_SRS`. Is there any way to share code? The difference seems mainly that the function takes a 512-bit src register and splits it, while in your case you already have two 256-bit regs (the two inputs to concat). 

How about changing the prototype of `select512BitG_AIE_STORE_SRS(..., Register SrcReg ...)` into `select512BitG_AIE_STORE_SRS(..., Register SrcRegLo, Register SrcRegHi ...)`?

The old code could be updated through something like
```
    auto [SrcRegLo, SrcRegHi] = split512BitReg(SrcReg);
    return select512BitG_AIE_STORE_SRS(*LSO, *AMI, SrcRegLo, SrcRegHi, ShftReg, SignReg, ConstantSign, MRI);
```